index.md

```diff
@@ -144,14 +142,28 @@ If this isn't sufficient for your design, please seek out experts to talk to.
+  * (EC)DSA have several different encodings: [ASN.1](https://en.wikipedia.org/wiki/Abstract_Syntax_Notation_One), IEEE P1363, and "raw" (my name). This means that an attacker can convert a valid signature in one form to the other.
+  * While both IEEE P1360 and raw have specific length requirements, many systems do no enforce them. (This is an example of an "edge-case" signature from later in the gotchas list.)
+  * While the ASN.1 encoding is supposed to be [DER](https://en.wikipedia.org/wiki/X.690#DER_encoding) encoded, many libraries accept any (semi-)valid [BER](https://en.wikipedia.org/wiki/X.690#BER_encoding) encoding. This means that an attacker can often use the flexibility of BER to craft an essentially infinite number of valid signatures (for the same message) once they know a single one. (more "edge-case" signatures)
   * ECDSA is mathematically malleable as well. It consists of two values `(r, s)` and given one signature it is trivial to calculate a new `s'` (equal to the order minus the original `s`) which results in a new valid signature `(r, s')` over the same message.
 * An attacker mustn't be allowed to select the actual value being validated in the signature. (The hashing step in all standard signatures defends against that as they can only select the hash pre-image, not the value of the hash.) If they could then they could trivially craft a valid signature for an arbitrary public key by (essentially) generating a random signature and seeing what message *would* be verified by that signature and then returning that message/signature pair.
 * Just because a signature is valid for a given message doesn't mean it isn't valid for other messages.
+  * For example in [EdDSA](https://en.wikipedia.org/wiki/EdDSA), an attacker can craft a malicious low-order public key and then create a signature valid for *all* messages.
+    (This, along with other properties of low-order points, was used to break [Scuttlebutt](https://en.wikipedia.org/wiki/Secure_Scuttlebutt) in section 7.1 of [this paper by Cremers and Jackson](https://eprint.iacr.org/2019/526) from 2020.)
+  * More trivially, signatures are generally over message digests, so if the hash function is broken (such as [MD5](https://eprint.iacr.org/2013/170) or [SHA-1](https://shattered.io/)) then a single signature is also valid for all messages with colliding hashes. [This was used to attack Windows Update in 2012](https://arstechnica.com/information-technology/2012/06/flame-crypto-breakthrough/).
+* Just because a signature is valid for a message and public key doesn't mean that the holder of the private key knows what the message was.
+  * Consider both the universal signature and hash collisions from an earlier point where the signatures are (by definition) valid for an unbounded collection of messages.
+  * There is also the trivial case that the signer might only know the hash and not the real message.
```

Comment:
Could add: "Some signature schemes (e.g. RSA-PKCSv1.5) allow an attacker to extract the message hash from a signature".

Reply:
I already have this:
> * Signatures do not (necessarily) hide the message they sign. They'll commonly reveal the hash (or more) of the data being signed.

Reply:
Ah sorry, I only looked at the diff.